### PR TITLE
Check block hash on old receipts

### LINKED ISSUE #490: Check block hash in transaction receipts
There's a small race condition in how we load transaction receipts for a block (see `EthereumAdapter::block_by_hash`). If the block being loaded is not part of the main chain of the Ethereum node (either because the block was removed due to a reorg, or because the block was added in a reorg that the particular Ethereum node has not yet seen), then the receipts returned may be from the wrong block. We should check that the block hash in the receipt matches the block's hash. Unfortunately, there does not seem to be a way to get transaction receipts by block.

### PR DESCRIPTION
**Pull request by timmclean**
There's a small chance of chain reorgs here, so make sure that receipts are for the right block.

Also, limit the number of retries due to a missing transaction receipt so that graph-node doesn't get stuck if the transaction never makes it back into the main chain.

Resolves #490

### CODE REVIEW

**Review by leoyvens — DISMISSED** [review]
Looks great, left a suggestion for refactoring.

File: datasource/ethereum/src/ethereum_adapter.rs [hunk 1]
```diff
@@ -295,7 +295,7 @@ where
         });
 
         let web3 = self.web3.clone();
-        Box::new(block_future.and_then(move |block_opt| {
+        Box::new(block_opt_future.and_then(move |block_opt| {
             let web3 = web3.clone();
 
             block_opt.map(move |block| {
```
[the remark this answers is not in the document]
Comment (timmclean) [reply datasource/ethereum/src/ethereum_adapter.rs]: Oh cool!! Forgot you could do that with Options

[the remark this answers is not in the document]
Comment (timmclean) [reply datasource/ethereum/src/ethereum_adapter.rs]: Ah wait... there's some type magic going on here to convert the `Option<impl Future>` into an `impl Future<Option>`. Is there a way to do it? I think I'd have to Box or something

[the remark this answers is not in the document]
Comment (timmclean) [reply datasource/ethereum/src/ethereum_adapter.rs]: Yeah note the `and_then`

[the remark this answers is not in the document]
Comment (leodasvacas) [reply datasource/ethereum/src/ethereum_adapter.rs]: I'm surprised that `?` won't work, it's not being easy you can just use a match with `None => return None`. But maybe I'm missing something and we should just leave it as is if this starts taking your time.

**Review by leoyvens — APPROVED** [review]

**Pull request merged**